File: CODIGOS5/173.rb
def determine_chord_name(notes)
    steps = []
    (1...notes.length).each do |i|
      diff = (notes[i] - notes[i - 1] + 12) % 12
      steps << diff
    end
  
    return 'other' if steps.include?(3) && steps.include?(4)
  
    if steps.include?(4)
      return get_chord_name(notes[0], true)
    elsif steps.include?(3)
      return get_chord_name(notes[0], false)
    else
      return 'other'
    end
  end
  
  def get_chord_name(root_note, is_major)
    note_names = %w[C C# D D# E F F# G G# A A# B]
    root_note_name = note_names[root_note % 12]
    return "#{root_note_name}-major" if is_major
  
    "#{root_note_name}-minor"
  end
  
  def contains(arr, target)
    arr.include?(target)
  end
  
  def main
    num_test_cases = gets.to_i
  
    (1..num_test_cases).each do
      input_notes = gets.chomp.split.map(&:to_i)
      chord_name = determine_chord_name(input_notes)
      print "#{chord_name} "
    end
  end
  
  main
  

File: CODIGOS5/173.pl
sub determineChordName {
    my @notes = @_;
    my @steps;
    for (my $i = 1; $i < scalar(@notes); $i++) {
        my $diff = ($notes[$i] - $notes[$i - 1] + 12) % 12;
        push @steps, $diff;
    }

    if (contains(\@steps, 3) && contains(\@steps, 4)) {
        return "other";
    } elsif (contains(\@steps, 4)) {
        return getChordName($notes[0], 1);
    } elsif (contains(\@steps, 3)) {
        return getChordName($notes[0], 0);
    } else {
        return "other";
    }
}

sub getChordName {
    my ($rootNote, $isMajor) = @_;
    my @noteNames = qw(C C# D D# E F F# G G# A A# B);
    my $rootNoteName = $noteNames[$rootNote % 12];
    if ($isMajor) {
        return "$rootNoteName-major";
    }
    return "$rootNoteName-minor";
}

sub contains {
    my ($arrRef, $target) = @_;
    foreach my $num (@$arrRef) {
        return 1 if $num == $target;
    }
    return 0;
}

sub main {
    my $numTestCases = <>;
    chomp($numTestCases);

    for (my $i = 0; $i < $numTestCases; $i++) {
        my $inputNotes = <>;
        chomp($inputNotes);
        my @notes = split(' ', $inputNotes);

        my $chordName = determineChordName(@notes);
        print "$chordName ";
    }
}

main();
